panther_diagnostics/src/system_status_node.cpp

```diff
@@ -0,0 +1,174 @@
+
+#include "panther_diagnostics/system_status_node.hpp"
+
+#include <chrono>
+#include <exception>
+#include <filesystem>
+#include <fstream>
+
+#include "diagnostic_updater/diagnostic_updater.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+#include "cppuprofile/uprofile.h"
+#include "panther_msgs/msg/system_status.hpp"
+
+namespace panther_diagnostics
+{
+
+SystemStatusNode::SystemStatusNode(const std::string & node_name)
+: rclcpp::Node(node_name), diagnostic_updater_(this)
+{
+  param_listener_ =
+    std::make_shared<system_status::ParamListener>(this->get_node_parameters_interface());
+  params_ = param_listener_->get_params();
+
+  system_status_publisher_ = this->create_publisher<panther_msgs::msg::SystemStatus>(
+    "system_status", 10);
+  timer_ = this->create_wall_timer(
+    std::chrono::milliseconds(static_cast<long long>(params_.publish_rate * 1000)),
+    std::bind(&SystemStatusNode::TimerCallback, this));
+
+  diagnostic_updater_.setHardwareID(params_.hardware_id);
+  diagnostic_updater_.add("OS status", this, &SystemStatusNode::DiagnoseSystem);
+
+  RCLCPP_INFO(this->get_logger(), "Node started");
+}
+
+void SystemStatusNode::TimerCallback()
+{
+  const auto status = GetSystemStatus();
+  auto message = BuildSystemStatusMessageFromSystemStatus(status);
+
+  system_status_publisher_->publish(message);
+}
+
+float SystemStatusNode::GetCoreTemperature(const std::string & filename) const
+{
+  try {
+    auto file = OpenFile(filename, std::ios_base::in);
+    float temperature;
+    file >> temperature;
+    file.close();
+    return temperature / 1000.0;
+  } catch (const std::runtime_error & e) {
+    const std::string msg = std::string("Error when trying to CPU temperature: ") + e.what();
+    RCLCPP_ERROR_STREAM(this->get_logger(), msg);
+  }
+  return std::numeric_limits<float>::quiet_NaN();
+}
+
+std::vector<float> SystemStatusNode::GetCoresUsages() const
+{
+  std::vector<float> loads = uprofile::getInstantCpuUsage();
+  return loads;
+}
+
+float SystemStatusNode::GetCoreMeanUsage(const std::vector<float> & usages) const
+{
+  auto sum = std::accumulate(usages.begin(), usages.end(), 0.0);
+  return sum / usages.size();
+}
+
+float SystemStatusNode::GetDiskUsage() const
+{
+  const std::filesystem::directory_entry entry("/");
+  const std::filesystem::space_info si = std::filesystem::space(entry.path());
+
+  return static_cast<float>(si.capacity - si.available) / si.capacity * 100.0;
+}
+
+float SystemStatusNode::GetMemoryUsage() const
+{
+  int total = 0, free = 0, available = 0;
+  uprofile::getSystemMemory(total, free, available);
+  return static_cast<float>(total - available) / total * 100.0;
+}
+
+SystemStatusNode::SystemStatus SystemStatusNode::GetSystemStatus() const
+{
+  SystemStatusNode::SystemStatus status;
+
+  status.core_usages_ = GetCoresUsages();
+  status.mean_core_usage_ = GetCoreMeanUsage(status.core_usages_);
+  status.core_temperature_ = GetCoreTemperature(SystemStatusNode::kTemperatureInfoFilename);
```

Comment:
Scope resolution is not needed here

```suggestion
  status.core_temperature_ = GetCoreTemperature(kTemperatureInfoFilename);
```